### PR TITLE
feat(metrics): add jvm class loader metrics

### PR DESCRIPTION
**Pull request by sliesche**

### CODE REVIEW

File: sda-commons-server-prometheus/src/main/java/org/sdase/commons/server/prometheus/PrometheusBundle.java [hunk 1]
```diff
@@ -116,6 +119,7 @@ private static void bindJvmAndSystemMetricsToGlobalRegistry() {
     new JvmGcMetrics().bindTo(Metrics.globalRegistry);
     new ProcessorMetrics().bindTo(Metrics.globalRegistry);
     new JvmThreadMetrics().bindTo(Metrics.globalRegistry);
+    new ClassLoaderMetrics().bindTo(Metrics.globalRegistry);
```
Comment (christopher-cudennec) [inline sda-commons-server-prometheus/src/main/java/org/sdase/commons/server/prometheus/PrometheusBundle.java:122]: I wonder if we need to close all metrics as described [here](https://github.com/micrometer-metrics/micrometer/issues/1492). Should we create another subtask for that story?

**Review by christopher-cudennec — COMMENTED** [review]
Do you usually add a test for our metrics?

File: sda-commons-server-prometheus/src/main/java/org/sdase/commons/server/prometheus/PrometheusBundle.java (hunk 1 above)
Comment (sliesche) [reply sda-commons-server-prometheus/src/main/java/org/sdase/commons/server/prometheus/PrometheusBundle.java:119]: The JvmGcMetrics are the only ones that are auto closable.
I added the shutdown hook in this PR using a separate commit.

Please let me know if you prefer a separate PR

Comment (christopher-cudennec) [reply sda-commons-server-prometheus/src/main/java/org/sdase/commons/server/prometheus/PrometheusBundle.java:119]: That's fine. Thanks for the quick fix! 👍

**Comment by sliesche** [comment]
> Do you usually add a test for our metrics?

Tests added

File: sda-commons-server-prometheus/src/test/java/org/sdase/commons/server/prometheus/PrometheusBundleTest.java [hunk 2]
```diff
@@ -240,6 +242,46 @@ void micrometerMetricsAvailableInPrometheus2() {
...
+        meters.stream()
+            .filter(m -> m.getId().getName().startsWith("jvm.gc"))
+            .collect(Collectors.toList());
+
+    assertThat(jvmBufferList).hasSize(6);
+    assertThat(jvmMemoryList).hasSize(24);
+    assertThat(jvmProcessorList).hasSize(3);
+    assertThat(jvmThreadsList).hasSize(10);
+    assertThat(jvmClassesList).hasSize(2);
+    assertThat(jvmGcList).hasSize(5);
```
Comment (christopher-cudennec) [inline sda-commons-server-prometheus/src/test/java/org/sdase/commons/server/prometheus/PrometheusBundleTest.java:282]: Tests are failing now for Java 17. Maybe too strict / number of meters depend on the Java version?

Comment (sliesche) [reply sda-commons-server-prometheus/src/test/java/org/sdase/commons/server/prometheus/PrometheusBundleTest.java]: Yes, metrics depend on Java version :worried:
I've made the tests less strict

File: sda-commons-server-prometheus/src/test/java/org/sdase/commons/server/prometheus/PrometheusBundleTest.java [hunk 3]
```diff
@@ -240,6 +242,46 @@ void micrometerMetricsAvailableInPrometheus2() {
...
+        meters.stream()
+            .filter(m -> m.getId().getName().startsWith("jvm.gc"))
+            .collect(Collectors.toList());
+
+    assertThat(jvmBufferList).hasSizeGreaterThanOrEqualTo(6);
+    assertThat(jvmMemoryList).hasSizeGreaterThanOrEqualTo(24);
+    assertThat(jvmProcessorList).hasSizeGreaterThanOrEqualTo(3);
+    assertThat(jvmThreadsList).hasSizeGreaterThanOrEqualTo(10);
+    assertThat(jvmClassesList).hasSizeGreaterThanOrEqualTo(2);
+    assertThat(jvmGcList).hasSizeGreaterThanOrEqualTo(5);
```
Comment (christopher-cudennec) [inline sda-commons-server-prometheus/src/test/java/org/sdase/commons/server/prometheus/PrometheusBundleTest.java:282]: Should we add a comment why we use ">="?

**Review by christopher-cudennec — APPROVED** [review]

**Review by christopher-cudennec — APPROVED** [review]

**Pull request merged**